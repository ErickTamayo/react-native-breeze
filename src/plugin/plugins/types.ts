import { ImageStyle, StyleProp, TextStyle, ViewStyle } from "react-native";
import Config from "../utils/config";

type PatterCallableParams = {
  keys: typeof Config.keys;
};

export type PatternCallable = (params: PatterCallableParams) => RegExp;

export type PluginPattern = RegExp | PatternCallable;

export type PluginFunctionParams<T> = {
  input: string;
  groups: T;
  theme: typeof Config.theme;
  color: typeof Config.color;
};

export type PluginFunctionReturnType =
  | StyleProp<ViewStyle>
  | StyleProp<ImageStyle>
  | StyleProp<TextStyle>;

export type PluginFunction<T = {}> = (
  params: PluginFunctionParams<T>
) => PluginFunctionReturnType;
