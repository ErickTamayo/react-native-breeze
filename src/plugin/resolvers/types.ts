export type ResolverFunctionParams = {
  input: string;
  groups: { [group: string]: string };
  theme: (key: string) => any;
};

export type ResolverFunction = (params: ResolverFunctionParams) => object;

export type Resolver = {
  pattern: RegExp;
  fn: ResolverFunction;
};
