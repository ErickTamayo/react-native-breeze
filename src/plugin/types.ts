export type Configuration = {
  theme: {
    [key: string]: any;
  };
  variants: {
    [key: string]: any;
  };
  corePlugins: {
    [key: string]: boolean;
  };
  plugins: string[];
};
