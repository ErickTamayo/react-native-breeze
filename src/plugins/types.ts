import { ImageStyle, StyleProp, TextStyle, ViewStyle } from "react-native";
import { keys, colorKeys, theme, color } from "../config";

type PatterCallableParams = {
  keys: typeof keys;
  colorKeys: typeof colorKeys;
};

export type PatternCallable = (params: PatterCallableParams) => RegExp;

export type PluginPattern = RegExp | PatternCallable;

export type PluginFunctionParams<T> = {
  input: string;
  groups: T;
  theme: typeof theme;
  color: typeof color;
};

export type PluginFunctionReturnType =
  | StyleProp<ViewStyle>
  | StyleProp<ImageStyle>
  | StyleProp<TextStyle>;

export type PluginFunction<T = {}> = (
  params: PluginFunctionParams<T>
) => PluginFunctionReturnType;
