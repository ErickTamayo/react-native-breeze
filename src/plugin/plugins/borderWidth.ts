import { validate } from "../utils/validation";
import { PluginFunction, PluginPattern } from "./types";

export type PluginGroups = {
  position?:
    | "t"
    | "r"
    | "b"
    | "l"
    | "s"
    | "e"
    | "tl"
    | "tr"
    | "ts"
    | "te"
    | "br"
    | "bl"
    | "bs"
    | "be";
  width?: string;
};

export const pattern: PluginPattern = ({ keys }) => {
  // prettier-ignore
  return new RegExp(`^border-?(?<position>t|r|b|l|s|e|tl|tr|ts|te|br|bl|bs|be)?-?(?<width>${keys("borderWidth")})?$`);
};

export const plugin: PluginFunction<PluginGroups> = ({
  input,
  groups,
  theme,
}) => {
  const { position, width } = groups;
  const value = theme<number>(["borderWidth", width || "default"]);

  if (!validate(input, value, ["number"])) return {};

  switch (position) {
    case "t":
      return { borderTopWidth: value };
    case "r":
      return { borderRightWidth: value };
    case "b":
      return { borderBottomWidth: value };
    case "l":
      return { borderLeftWidth: value };
    case "s":
      return { borderStartWidth: value };
    case "e":
      return { borderEndWidth: value };
    case "tl":
      return { borderTopLeftWidth: value };
    case "tr":
      return { borderTopRightWidth: value };
    case "ts":
      return { borderTopStartWidth: value };
    case "te":
      return { borderTopEndWidth: value };
    case "bl":
      return { borderBottomLeftWidth: value };
    case "br":
      return { borderBottomRightWidth: value };
    case "bs":
      return { borderBottomStartWidth: value };
    case "be":
      return { borderBottomEndWidth: value };
    default:
      return { borderWidth: value };
  }
};
