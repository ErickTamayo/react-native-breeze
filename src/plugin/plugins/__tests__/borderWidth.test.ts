import { pattern, plugin, PluginGroups } from "../borderWidth";
import { PatternCallable, PluginFunctionReturnType } from "../types";

describe("borderWidth", () => {
  it("Should match the correct pattern", () => {
    const theme = jest.fn().mockReturnValue({ 2: 2, default: 1 });
    const callable = pattern as PatternCallable;

    expect(callable({ theme }).exec("border")).toBeTruthy();
    expect(callable({ theme }).exec("border-2")).toBeTruthy();

    expect(callable({ theme }).exec("border-t")).toBeTruthy();
    expect(callable({ theme }).exec("border-r")).toBeTruthy();
    expect(callable({ theme }).exec("border-b")).toBeTruthy();
    expect(callable({ theme }).exec("border-l")).toBeTruthy();
    expect(callable({ theme }).exec("border-s")).toBeTruthy();
    expect(callable({ theme }).exec("border-e")).toBeTruthy();
    expect(callable({ theme }).exec("border-tl")).toBeTruthy();
    expect(callable({ theme }).exec("border-tr")).toBeTruthy();
    expect(callable({ theme }).exec("border-ts")).toBeTruthy();
    expect(callable({ theme }).exec("border-te")).toBeTruthy();
    expect(callable({ theme }).exec("border-br")).toBeTruthy();
    expect(callable({ theme }).exec("border-bl")).toBeTruthy();
    expect(callable({ theme }).exec("border-bs")).toBeTruthy();
    expect(callable({ theme }).exec("border-be")).toBeTruthy();

    expect(callable({ theme }).exec("border-t-2")).toBeTruthy();
    expect(callable({ theme }).exec("border-r-2")).toBeTruthy();
    expect(callable({ theme }).exec("border-b-2")).toBeTruthy();
    expect(callable({ theme }).exec("border-l-2")).toBeTruthy();
    expect(callable({ theme }).exec("border-s-2")).toBeTruthy();
    expect(callable({ theme }).exec("border-e-2")).toBeTruthy();
    expect(callable({ theme }).exec("border-tl-2")).toBeTruthy();
    expect(callable({ theme }).exec("border-tr-2")).toBeTruthy();
    expect(callable({ theme }).exec("border-ts-2")).toBeTruthy();
    expect(callable({ theme }).exec("border-te-2")).toBeTruthy();
    expect(callable({ theme }).exec("border-br-2")).toBeTruthy();
    expect(callable({ theme }).exec("border-bl-2")).toBeTruthy();
    expect(callable({ theme }).exec("border-bs-2")).toBeTruthy();
    expect(callable({ theme }).exec("border-be-2")).toBeTruthy();

    expect(callable({ theme }).exec("border-unknown")).toBeFalsy();
    expect(callable({ theme }).exec("border-t-unknown")).toBeFalsy();
    expect(callable({ theme }).exec("border-r-unknown")).toBeFalsy();
    expect(callable({ theme }).exec("border-b-unknown")).toBeFalsy();
    expect(callable({ theme }).exec("border-l-unknown")).toBeFalsy();
    expect(callable({ theme }).exec("border-tl-unknown")).toBeFalsy();
    expect(callable({ theme }).exec("border-tr-unknown")).toBeFalsy();
    expect(callable({ theme }).exec("border-br-unknown")).toBeFalsy();
    expect(callable({ theme }).exec("border-bl-unknown")).toBeFalsy();
  });

  // prettier-ignore
  it.each<[input: string, width: { [key: string]: number }, returned: number, expected: PluginFunctionReturnType]>([
    ["border", { default: 1, 2: 2 }, 4, { borderWidth: 4 }],
    ["border-t", { default: 1, 2: 2 }, 4, { borderTopWidth: 4 } as any], //TODO check why this typing is failing
    ["border-r", { default: 1, 2: 2 }, 4, { borderRightWidth: 4 }],
    ["border-b", { default: 1, 2: 2 }, 4, { borderBottomWidth: 4 }],
    ["border-l", { default: 1, 2: 2 }, 4, { borderLeftWidth: 4 }],
    ["border-s", { default: 1, 2: 2 }, 4, { borderStartWidth: 4 }],
    ["border-e", { default: 1, 2: 2 }, 4, { borderEndWidth: 4 }],
    ["border-tl", { default: 1, 2: 2 }, 4, { borderTopLeftWidth: 4 }],
    ["border-tr", { default: 1, 2: 2 }, 4, { borderTopRightWidth: 4 }],
    ["border-br", { default: 1, 2: 2 }, 4, { borderBottomRightWidth: 4 }],
    ["border-bl", { default: 1, 2: 2 }, 4, { borderBottomLeftWidth: 4 }],
    ["border-2", { default: 1, 2: 2 }, 2, { borderWidth: 2 }],
    ["border-t-2", { default: 1, 2: 2 }, 2, { borderTopWidth: 2 } as any], //TODO check why this typing is failing
    ["border-r-2", { default: 1, 2: 2 }, 2, { borderRightWidth: 2 }],
    ["border-b-2", { default: 1, 2: 2 }, 2, { borderBottomWidth: 2 }],
    ["border-l-2", { default: 1, 2: 2 }, 2, { borderLeftWidth: 2 }],
    ["border-s-2", { default: 1, 2: 2 }, 2, { borderStartWidth: 2 }],
    ["border-e-2", { default: 1, 2: 2 }, 2, { borderEndWidth: 2 }],
    ["border-tl-2", { default: 1, 2: 2 }, 2, { borderTopLeftWidth: 2 }],
    ["border-tr-2", { default: 1, 2: 2 }, 2, { borderTopRightWidth: 2 }],
    ["border-br-2", { default: 1, 2: 2 }, 2, { borderBottomRightWidth: 2 }],
    ["border-bl-2", { default: 1, 2: 2 }, 2, { borderBottomLeftWidth: 2 }],
   ])(
    "Should parse (%s) correctly",
    (input, radii, returned, expected) => {
      const theme = jest.fn()
      
      theme.mockReturnValueOnce(radii);
      const callable = pattern as PatternCallable;
      const groups = callable({ theme }).exec(input)!.groups! as PluginGroups;

      theme.mockReturnValueOnce(returned);
      const result = plugin({ input, groups, theme });

      expect(result).toEqual(expected);
    }
  );

  it("Should log error if the borderWidth value is not valid", () => {
    const theme = jest.fn();

    theme.mockReturnValueOnce({ default: "0" });
    const input = "border";
    const callable = pattern as PatternCallable;
    const groups = callable({ theme }).exec(input)!.groups! as PluginGroups;
    const spy = jest.spyOn(console, "error").mockImplementation(() => {});

    theme.mockReturnValueOnce("0");
    const result = plugin({ input, groups, theme });

    expect(result).toEqual({});
    expect(spy).toHaveBeenCalled();

    spy.mockRestore();
  });
});
