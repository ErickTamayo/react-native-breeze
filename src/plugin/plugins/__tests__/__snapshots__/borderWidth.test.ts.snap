// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`borderWidth Should parse (border) correctly 1`] = `
Object {
  "borderWidth": 1,
}
`;

exports[`borderWidth Should parse (border-2) correctly 1`] = `
Object {
  "borderWidth": 2,
}
`;

exports[`borderWidth Should parse (border-b) correctly 1`] = `
Object {
  "borderBottomWidth": 1,
}
`;

exports[`borderWidth Should parse (border-b-2) correctly 1`] = `
Object {
  "borderBottomWidth": 2,
}
`;

exports[`borderWidth Should parse (border-be) correctly 1`] = `
Object {
  "borderBottomEndWidth": 1,
}
`;

exports[`borderWidth Should parse (border-be-2) correctly 1`] = `
Object {
  "borderBottomEndWidth": 2,
}
`;

exports[`borderWidth Should parse (border-bl) correctly 1`] = `
Object {
  "borderBottomLeftWidth": 1,
}
`;

exports[`borderWidth Should parse (border-bl-2) correctly 1`] = `
Object {
  "borderBottomLeftWidth": 2,
}
`;

exports[`borderWidth Should parse (border-br) correctly 1`] = `
Object {
  "borderBottomRightWidth": 1,
}
`;

exports[`borderWidth Should parse (border-br-2) correctly 1`] = `
Object {
  "borderBottomRightWidth": 2,
}
`;

exports[`borderWidth Should parse (border-bs) correctly 1`] = `
Object {
  "borderBottomStartWidth": 1,
}
`;

exports[`borderWidth Should parse (border-bs-2) correctly 1`] = `
Object {
  "borderBottomStartWidth": 2,
}
`;

exports[`borderWidth Should parse (border-e) correctly 1`] = `
Object {
  "borderEndWidth": 1,
}
`;

exports[`borderWidth Should parse (border-e-2) correctly 1`] = `
Object {
  "borderEndWidth": 2,
}
`;

exports[`borderWidth Should parse (border-l) correctly 1`] = `
Object {
  "borderLeftWidth": 1,
}
`;

exports[`borderWidth Should parse (border-l-2) correctly 1`] = `
Object {
  "borderLeftWidth": 2,
}
`;

exports[`borderWidth Should parse (border-r) correctly 1`] = `
Object {
  "borderRightWidth": 1,
}
`;

exports[`borderWidth Should parse (border-r-2) correctly 1`] = `
Object {
  "borderRightWidth": 2,
}
`;

exports[`borderWidth Should parse (border-s) correctly 1`] = `
Object {
  "borderStartWidth": 1,
}
`;

exports[`borderWidth Should parse (border-s-2) correctly 1`] = `
Object {
  "borderStartWidth": 2,
}
`;

exports[`borderWidth Should parse (border-t) correctly 1`] = `
Object {
  "borderTopWidth": 1,
}
`;

exports[`borderWidth Should parse (border-t-2) correctly 1`] = `
Object {
  "borderTopWidth": 2,
}
`;

exports[`borderWidth Should parse (border-te) correctly 1`] = `
Object {
  "borderTopEndWidth": 1,
}
`;

exports[`borderWidth Should parse (border-te-2) correctly 1`] = `
Object {
  "borderTopEndWidth": 2,
}
`;

exports[`borderWidth Should parse (border-tl) correctly 1`] = `
Object {
  "borderTopLeftWidth": 1,
}
`;

exports[`borderWidth Should parse (border-tl-2) correctly 1`] = `
Object {
  "borderTopLeftWidth": 2,
}
`;

exports[`borderWidth Should parse (border-tr) correctly 1`] = `
Object {
  "borderTopRightWidth": 1,
}
`;

exports[`borderWidth Should parse (border-tr-2) correctly 1`] = `
Object {
  "borderTopRightWidth": 2,
}
`;

exports[`borderWidth Should parse (border-ts) correctly 1`] = `
Object {
  "borderTopStartWidth": 1,
}
`;

exports[`borderWidth Should parse (border-ts-2) correctly 1`] = `
Object {
  "borderTopStartWidth": 2,
}
`;
